### PR TITLE
feat(DataTable): add support for alphanumeric, datatable, and custom sort functions

### PR DESCRIPTION
**Pull request by joshblack**
Closes https://github.com/github/primer/issues/1832
Closes https://github.com/github/primer/issues/1833

#### Changelog

**New**

- Add `alphanumeric` sort strategy
- Add `datetime` sort strategy
- Add support for a `CustomSortStrategy`
- Add support for custom sort strategy

**Changed**

- Update stories to use `alphanumeric` and `datetime` sort strategies

**Removed**

### CODE REVIEW

**Comment by joshblack** [comment]
@colebemis if you have a second, would love your help on some of the TS issues I'm running into in the sort method 🤔 Really not sure what to do (it's the `todo` comments over in [this file](https://github.com/primer/react/pull/3001/files#diff-8eb04eebc0522f8aca89dc08182490363338a06c61e37a6d2846ada3eeeab5c3))

**Review by colebemis — APPROVED** [review]
Looks great! Love how extensible this API is 😍

File: src/DataTable/sorting.ts [hunk 1]
```diff
@@ -24,11 +24,112 @@ export function transition(direction: Exclude<SortDirection, 'NONE'>): Exclude<S
...
+    } else if (typeof a === 'number' && typeof b === 'number') {
+      // If both groups are numbers, compare them numerically
+      return a > b ? 1 : -1
+    } else if (typeof a === 'number' && typeof b === 'string') {
+      // Sort numbers before strings
+      return -1
+    } else if (typeof a === 'string' && typeof b === 'number') {
+      // Sort numbers before strings
+      return 1
+    } else if (a === undefined || b === undefined) {
```
Comment (mperrotti) [inline src/DataTable/sorting.ts:79]: What if `a` or `b` are `null`?

File: src/DataTable/useTable.ts [hunk 2]
```diff
@@ -168,13 +168,25 @@ export function useTable<Data extends UniqueRow>({
 
     setRowOrder(rowOrder => {
       return rowOrder.slice().sort((a, b) => {
+        // Custom sort functions operate on the row versus the field
+        if (typeof header.column.sortBy === 'function') {
+          if (state.direction === SortDirection.ASC) {
+            // @ts-ignore todo
```
Comment (mperrotti) [inline src/DataTable/useTable.ts:174]: I didn't know `@ts-ignore todo` was a thing!

**Review by mperrotti — APPROVED** [review]
Great stuff!

File: src/DataTable/sorting.ts (hunk 1 above)
Comment (joshblack) [reply src/DataTable/sorting.ts:79]: Great call out, I swear any `undefined` call site has me anxious with `null` too 😂 Thankfully `Array.prototype.shift()` should be `string | number | undefined` so we don't have to worry about `null` (I hope 🤞)

**Pull request merged**